### ISSUE TITLE
Create a source of truth for accessibility in JupyterHub

### ISSUE THREAD

**Issue by choldgraf** [jupyterhub/roadmap#12]
### Problem Statement

A few user stories to consider:

- As a learner or researcher who requires assistive devices to do my data science work, I need interfaces that have been designed for these devices to interact with. (one could imagine a similar story for the many other ways in which people need perceptual assistance working in data science)
- As an administrator at an organization, I have a mandate to use technology that meets minimal criteria for accessibility for the user communities I must serve. I need a simple way to answer "is this technology accessible?"

Currently, there is no single place that is discoverable and explains all of the information somebody needs to understand the current state of accessibility in JupyterHub, and where to go for more information. This means people waste time digging into jupyterhub docs, go down rabbit holes because they have the wrong mental models for what jupyterhub does and doesn't do, and might choose not to use (or not to use) JupyterHub because of incorrect assumptions based on accessibility.

This is becoming a bigger problem as a11y is increasingly mandated by policy at public institutions (e.g. teaching institutions). There are also some formal practices for defining accessibility that might make this easier at a policy level.

### Proposed Solution

Add a single "landing page" for JupyterHub to define it's high-level "accessibility story". Here are a few thoughts:

- Something that just speaks to whether our user interfaces (hub login etc) meet various definitions of accessibility. This would be a helpful resource for us to point to elsewhere, as users are increasingly interested in knowing they are using accessible tools.
- Explanation of what is in / out of scope for JupyterHub interfaces (e.g. we can't control the a11y of the interfaces you _serve_ with JupyterHub, like JupyterLab or VSCode, but we can control a11y of hub's interfaces like the login screen)
- Ideally, a link to a VPAT for JupyterHub (ref: https://github.com/jupyter/accessibility/issues/155) but doesn't need to block this
tudio / VSCode / MATLAB / Positron.
- Links to the other accessibility spaces in Jupyter to learn more, ref: https://github.com/jupyter/accessibility/issues/154

### Proposed Implementation

I'd imagined any of the following places, and with minimal information and/or links to learn more deeply (e.g., a link to an accessibility github issue tag etc):

- A page tied to jupyterhub the app (e.g. [jupyterhub.readthedocs.io/en/latest/accessibility](http://jupyterhub.readthedocs.io/en/latest/)
- A page tied to the jupyterhub _project_ as a whole (e.g., [hub.jupyter.org/accessibility](http://hub.jupyter.org/)
- A page on [jupyter.org](http://jupyter.org/) (e.g. [jupyter.org/hub/accessibility](http://jupyter.org/hub/accessibility))

Links to ongoing work to check and improve this

- https://github.com/jupyterhub/jupyterhub/issues/4078

### Endorsements

-

**Comment by minrk** [comment]
_If_ we decide to populate a 'project' landing side (hub.jupyter.org), I agree this would belong there. Since we haven't really done that yet, we could either:

1. use this as an opportunity to do that!
2. put it in jupyterhub docs for now as the logical central location. the jupyterhub package does contain _almost_ all of the JuptyerHub UI (except for fancy profiles, pretty much, which 2i2c owns, not jupyterhub)

I'd _love_ help on VPAT / accessibility testing. @Bl-aire made a number of excellent accessibility improvements to JupyterHub pages, but we don't have accessibility _testing_ to help us not backslide.

JupyterLab really is the biggest project that matters, since students hardly see JupyterHub UI. But that also means we should be able to easily meet high standards!

**Comment by minrk** [comment]
btw, I just learned about https://github.com/pamelafox/axe-playwright-python which was presented two years ago at https://northbaypython.org a truly excellent local event I was at just 3 days ago!

**Comment by bl-aire** [comment]
Hello @minrk @choldgraf,

With your guidance, I would like to work on this.
I can start by researching the current state of accessibility in JupyterHub and putting together a draft page along with some proposed next steps. The draft could include outlining what JupyterHub covers, what’s out of scope, and identifying gaps (like testing or VPAT). From there, I can open a PR or share a proposal for feedback.

Let me know if that sounds like a good approach.

**Comment by choldgraf** [comment]
IMO we definitely don't need anything too complex, because I agree with @minrk 's take here:

> JupyterLab really is the biggest project that matters, since students hardly see JupyterHub UI. But that also means we should be able to easily meet high standards!

But I think there's still a lot of value in "saying this" as the JupyterHub project so that others can discover it and don't need to figure it out for themselves. We should also clarify that for any interface, JupyterHub doesn't have the ability to control a11y for it (e.g. if somebody runs VSCode on JupyterHub, a11y in VSCode isn't in hub's control. This is a pretty common point of confusion)

@bl-aire I would personally find that really helpful!

**Comment by minrk** [comment]
@bl-aire that would be great! Let me know if you need any support.

**Comment by bl-aire** [comment]
I will get to work then!
Thank you @minrk @choldgraf

**Comment by bl-aire** [comment]
Hello @minrk,

While setting up a development install of JupyterHub, I ran into a blocker that seemed related to event schema validation.

When I ran: 
JupyterHub -f testing/jupyterhub_config.py

I got the following error:
jsonschema.exceptions.ValidationError: “1” is not of type “integer”

From the traceback, this appears to originate in the event logging setup via jupyter_events, where the schema defines "version": "1" (string) but the validator expects an integer.

I was able to work around it locally by changing version: “1” to version: 1 in event-schemas/v1.yaml but I wanted to confirm:

1. Is this a known issue with the dev branch?
2. Would you prefer a PR to update the schema ("version": 1) or that can come later?

**Comment by minrk** [comment]
Make sure to upgrade `jupyter-events`, the validator should expect a string. This changed in jupyter-events 0.11:

```
pip install --upgrade jupyter-events
```

**Comment by bl-aire** [comment]
> Make sure to upgrade `jupyter-events`, the validator should expect a string. This changed in jupyter-events 0.11:
> 
> ```
> pip install --upgrade jupyter-events
> ```

Okay, I will.
Thanks

**Comment by bl-aire** [comment]
Hello @minrk,

I’ve created and shared an initial draft and added you as an editor on the Google document.

I’d appreciate any feedback or suggestions you may have.

**Comment by bl-aire** [comment]
Hi @minrk,

I thought to follow up on the draft I shared. I'd love to get your thoughts on the direction, structure, or anything you think could be improved. I could also paste the link here for easier access.
Thank you

**Comment by minrk** [comment]
Thanks for the reminder! Sorry, it had slipped off my radar. I think the draft has lots of good things in it, but it's unclear to me what an appropriate level of detail is. That document is _very_ detailed. What did you use as a reference to put it together? Do we have good examples to base ours on?

It seems like most examples like:

- https://accessible.org/examples-vpats-acrs/
- https://appian.com/legal/accessibility-conformance-report-vpat-508
- https://www.mediawiki.org/wiki/VPAT_for_MediaWiki
- https://website-archive.mozilla.org/www.mozilla.org/firefox_vpat/firefox-vpat-3.html
- https://accessibility.github.com/conformance

have far less narrative and more specific, bullet/tabular presentation of key areas where things have been evaluated. It seems like for our purposes, a much more terse document addressing the points @choldgraf mentioned, and an indication of the scope of the wcag testing that you did earlier would perhaps be more inline with what's expected?

**Comment by choldgraf** [comment]
As an example here is the accessibility page we have with the myst project: 

https://mystmd.org/guide/accessibility-and-performance

It's intentionally fairly short, and not not meant to be a full reference.

And the subproject has an even more high level accessibility page that points to the myst page

https://jupyterbook.org/accessibility/

**Comment by bl-aire** [comment]
Thanks for the thoughtful feedback @minrk @choldgraf .

The initial draft is not based on any single template, but rather, it is pulled from a mix of guidance around WCAG conformance documentation and some narrative-style reports I’ve seen in other contexts. 

Looking more closely at the examples you shared, I agree that most VPATs/ACRs tend to be much more structured and concise, with a stronger emphasis on tabular/bullet-based reporting rather than detailed narrative.
It would make sense to shift our approach toward a more standard VPAT-style format:
* Focus on clearly scoped WCAG criteria
* Provide concise, structured evaluations (tables/bullets)
* Include a brief summary of testing methodology and scope (e.g., what WCAG levels were assessed, tools/manual testing used, etc.)
* Call out any notable gaps or partial conformance areas without extensive prose
We can still retain a short introductory section for context, but overall keep the document tighter and more aligned with expectations for accessibility conformance reports.

I will revise the draft in that direction; simplifying the narrative and restructuring it to mirror the examples more closely while incorporating the points @choldgraf mentioned.

**Comment by bl-aire** [comment]
Hello @minrk @choldgraf,

I have updated the accessibility conformance report based on the recent feedback to make it more concise and aligned with VPAT-style expectations. The new version reduces the narrative significantly, is closer in style to the examples, and is intended as a high-level overview.

Before I proceed, I would really appreciate a review, especially on whether the level of detail and structure is appropriate, and if anything critical is missing.
Link: https://docs.google.com/document/d/1PsrSG6PWnQJYeiQ2-cbhBBF6ImxGWsjNm7Yay01Wsy8/edit?usp=sharing

Thank you!

**Comment by bl-aire** [comment]
> Hello [@minrk](https://github.com/minrk) [@choldgraf](https://github.com/choldgraf),
> 
> I have updated the accessibility conformance report based on the recent feedback to make it more concise and aligned with VPAT-style expectations. The new version reduces the narrative significantly, is closer in style to the examples, and is intended as a high-level overview.
> 
> Before I proceed, I would really appreciate a review, especially on whether the level of detail and structure is appropriate, and if anything critical is missing. Link: https://docs.google.com/document/d/1PsrSG6PWnQJYeiQ2-cbhBBF6ImxGWsjNm7Yay01Wsy8/edit?usp=sharing
> 
> Thank you!

Hi @minrk @choldgraf,

I just wanted to kindly follow up on this and see if there are any updates or feedback on the revised accessibility conformance report.
I’d really appreciate your thoughts when you have a chance. Thank you!

**Comment by minrk** [comment]
Sorry for dropping this. I think that document looks okay, but I still don't feel like I have the information with which to evaluate whether that's what folks need.

The link that @choldgraf [gave](https://mystmd.org/guide/accessibility-and-performance) looks like a really nice model to follow.

**Comment by bl-aire** [comment]
> Sorry for dropping this. I think that document looks okay, but I still don't feel like I have the information with which to evaluate whether that's what folks need.
> 
> The link that [@choldgraf](https://github.com/choldgraf) [gave](https://mystmd.org/guide/accessibility-and-performance) looks like a really nice model to follow.

Thank you, @minrk. No worries, and I appreciate the feedback.

I’ll use the MyST accessibility and performance page as a model and revise the report to better align with the information and structure needed for evaluation. I’ll share an updated version once it’s ready.
